packages/iris-grid/src/IrisGridModel.ts

```diff
@@ -0,0 +1,502 @@
+
+type IrisGridModelEventMap = {
+  [E in IrisGridModelEventNames]: Event<E>;
+};
+
+/**
+ * Abstract class that extends the GridModel to have more functionality, like filtering and sorting.
+ * For use from IrisGrid.
+ * Provides some abstraction from the dh.Table and dh.TreeTable classes, so we can treat them somewhat the same.
+ * Note that it still uses dh.Column, dh.FilterCondition, dh.Sort, etc., still. Theoretically should abstract
+ * those out as well, so there's no dependency on IrisAPI at all, but it's a lot of work for no real gain at this time.
+ */
+abstract class IrisGridModel<
+  TEventMap extends Record<string, Event<string>> = Record<
+    string,
+    Event<string>
+  >,
+  TMode extends 'standard' | 'strict' = 'standard'
+> extends GridModel<TEventMap & IrisGridModelEventMap, TMode> {
+  static EVENT = Object.freeze({
+    UPDATED: 'UPDATED',
+    FORMATTER_UPDATED: 'FORMATTER_UPDATED',
+    REQUEST_FAILED: 'REQUEST_FAILED',
+    COLUMNS_CHANGED: 'COLUMNS_CHANGED',
+    TABLE_CHANGED: 'TABLE_CHANGED',
+    FILTERS_CHANGED: 'FILTERS_CHANGED',
+    SORTS_CHANGED: 'SORTS_CHANGED',
+    DISCONNECT: 'DISCONNECT',
+    RECONNECT: 'RECONNECT',
+    TOTALS_UPDATED: 'TOTALS_UPDATED',
+    PENDING_DATA_UPDATED: 'PENDING_DATA_UPDATED',
+  } as const);
+
+  constructor() {
+    super();
+
+    this.listenerCount = 0;
+  }
+
+  listenerCount: number;
+
+  // Pulled directly from event-target-shim implementation signature
+  // https://github.com/mysticatea/event-target-shim/blob/master/src/lib/event-target.ts#L99
+  // Using Parameters<GridModel['addEventListener']> doesn't work
+  addEventListener<T extends string & keyof TEventMap>(
+    type0: T,
+    callback0?: EventTarget.EventListener<this, TEventMap[T]> | null,
+    options0?: boolean | EventTarget.AddOptions
+  ): void {
+    super.addEventListener(type0, callback0 as never, options0 as never);
+
+    this.listenerCount += 1;
+    if (this.listenerCount === 1) {
+      this.startListening();
+    }
+  }
+
+  removeEventListener<T extends string & keyof TEventMap>(
+    type0: T,
+    callback0?: EventTarget.EventListener<this, TEventMap[T]> | null,
+    options0?: boolean | EventTarget.Options
+  ): void {
+    super.removeEventListener(type0, callback0 as never, options0 as never);
+
+    this.listenerCount -= 1;
+    if (this.listenerCount === 0) {
+      this.stopListening();
+    }
+  }
+
+  /**
+   * Function called when first listener is added.
+   * Override for implementation specific behaviour.
+   */
+  startListening(): void {
+    // no-op
+  }
+
+  /**
+   * Function called when last listener is removed.
+   * Override for implementation specific behaviour.
+   */
+  stopListening(): void {
+    // no-op
+  }
+
+  /**
+   * Gets the columns for this model
+   * @returns All columns in the model
+   */
+  get columns(): Column[] {
+    throw new Error('get columns not implemented');
+  }
```

Comment:
`IrisGridModel` should ultimately be an abstract class, then we don't provide an impl for these at all. Right now it's not clear which methods an implementing class needs to override (e.g. `MyIrisGridModel extends IrisGridModel` won't get an error for not overriding `get columns()` even though it's required)

Reply:
I did make the class abstract, but I'm not sure what methods should be abstract and which shouldn't

Reply:
Any of the methods that throw and do nothing else should be abstract